### ISSUE TITLE
Add WithTag(tag, bool) for Traces

### ISSUE THREAD

**Issue by DoctorGurke** [Facepunch/sbox-issues#2042]
### What it is?

With the old HitLayer stuff, we could do `.HitLayer(CollisionLayer.Water, !inWater)` to add a condition for specific layers within the trace builder. This is no longer the case since we fully rely on tags, which does not have such an overload.

### What should it be?

an overload for the WithTag etc methods on trace builders that takes a bool as a last parameter, alternatively just as an optional parameter.

**Comment by handsomematt** [comment]
Does Trace.WithoutTag( string tag ) not do what you want?

**Comment by DoctorGurke** [comment]
correct me if I'm misunderstanding something, but since that's for the Trace directly and not the Traceresult, it requires some messier code blocks like these

old code:
```csharp
var tr = Trace.Ray( start, ent ).UseHitboxes().HitLayer( CollisionLayer.Water, !inWater ).Run();
```
new code:
```csharp
var t = Trace.Ray( start, end ).UseHitboxes();

if( !inWater )
{
    t = t.WithTag( "water" );
}

var tr = t.Run();
```

**Comment by garrynewman** [comment]
I can see the logic, but there's 3 types of tag filters, must have all, must have any, must not have.. so it seems like adding this would be adding extra complication to it all.

**Comment by DoctorGurke** [comment]
Fair enough, I see the downsides to this. The new code seems messier to me but is technically just more clear than previously, just gotta get used to it I suppose

**Comment by Redhacker1** [comment]
@DoctorGurke an extra side note, you could do similar functionality with extensions if I am not mistaken
